src/librustc/ty/relate.rs

```diff
@@ -453,10 +453,11 @@ pub fn super_relate_tys<'a, 'gcx, 'tcx, R>(relation: &mut R,
             Ok(tcx.mk_slice(t))
         }
 
-        (&ty::TyTuple(as_), &ty::TyTuple(bs)) =>
+        (&ty::TyTuple(as_, a_defaulted), &ty::TyTuple(bs, b_defaulted)) =>
         {
             if as_.len() == bs.len() {
-                Ok(tcx.mk_tup(as_.iter().zip(bs).map(|(a, b)| relation.relate(a, b)))?)
+                let defaulted = a_defaulted || b_defaulted;
```

Comment:
Hmm, I wonder if this should be `a_defaulted && b_defaulted`? My thinking is this: if we are relating two `()` values, and one of them came directly from the user, then that type will continue to be `()` even when `!` fallback is changed. Probably it doesn't matter because in such a case the fallback won't actually trigger anyhow.